File: src/hermes/reports.ts
/**
 * #20 — Hermes Summary Reports (TypeScript Port)
 *
 * Generates a Markdown summary report from a list of event dicts.
 * Reports include credential counts, outcome breakdowns, credit
 * consumption, top failure reasons, and recommended next actions.
 * Saved to `hermes/reports/` with timestamped filenames.
 *
 * Ported from hermes/reports.py
 */

import fs from "node:fs";
import path from "node:path";

const REPORTS_DIR = path.join(process.cwd(), "hermes", "reports");

function ensureReportsDir(): void {
  if (!fs.existsSync(REPORTS_DIR)) {
    fs.mkdirSync(REPORTS_DIR, { recursive: true });
  }
}

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface ReportEvent {
  data?: {
    outcome?: string;
    creditsSpent?: number;
    credential?: string;
    email?: string;
    username?: string;
    error?: string;
    message?: string;
    [key: string]: unknown;
  };
  [key: string]: unknown;
}

export interface RunSummary {
  total: number;
  uniqueCredentials: number;
  successCount: number;
  failCount: number;
  requeueCount: number;
  successRate: number;
  totalCredits: number;
  outcomeBreakdown: Map<string, number>;
  topFailures: Array<[string, number]>;
  recommendedActions: string[];
  reportPath: string;
  markdown: string;
}

// ---------------------------------------------------------------------------
// Counter utility
// ---------------------------------------------------------------------------

class Counter<T extends string | number> {
  private readonly _counts = new Map<T, number>();

  increment(key: T): void {
    this._counts.set(key, (this._counts.get(key) ?? 0) + 1);
  }

  get(key: T): number {
    return this._counts.get(key) ?? 0;
  }

  entries(): Array<[T, number]> {
    return [...this._counts.entries()];
  }

  mostCommon(n?: number): Array<[T, number]> {
    const sorted = this.entries().sort((a, b) => b[1] - a[1]);
    return n ? sorted.slice(0, n) : sorted;
  }

  total(): number {
    let sum = 0;
    for (const v of this._counts.values()) sum += v;
    return sum;
  }
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Produce a Markdown report summarising the given events.
 *
 * The report is also written to `hermes/reports/report-{timestamp}.md`.
 */
export function generateRunSummary(events: ReportEvent[]): RunSummary {
  const total = events.length;
  const outcomes = new Counter<string>();
  const failureReasons = new Counter<string>();
  const credentialsSeen = new Set<string>();
  let totalCredits = 0;

  for (const ev of events) {
    const data = ev.data ?? {};
    const outcome = String(data.outcome ?? "unknown");
    outcomes.increment(outcome);

    const cred = data.credential ?? data.email ?? data.username;
    if (cred) credentialsSeen.add(String(cred));

    totalCredits += Number(data.creditsSpent ?? 0);

    if (outcome !== "success" && outcome !== "requeue") {
      const reason = String(data.error ?? data.message ?? outcome);
      failureReasons.increment(reason);
    }
  }

  const successCount = outcomes.get("success");
  const requeueCount = outcomes.get("requeue");
  const failCount = outcomes.entries()
    .filter(([o]) => o !== "success" && o !== "requeue")
    .reduce((sum, [, c]) => sum + c, 0);
  const successRate = total > 0 ? (successCount / total) * 100 : 0;

  const topFailures = failureReasons.mostCommon(5);

  // Build recommended actions
  const actions: string[] = [];
  if (successRate < 50) {
    actions.push("⚠️  Success rate below 50% — investigate top failure reasons urgently.");
  }
  if (topFailures.some(([r]) => /selector|dom/i.test(r))) {
    actions.push(
      "🔧  DOM/selector failures detected — review engine.ts locators against latest page screenshots."
    );
  }
  if (topFailures.some(([r]) => /proxy|tunnel/i.test(r))) {
    actions.push(
      "🌐  Infrastructure failures detected — rotate proxy endpoints or check tunnel health."
    );
  }
  if (topFailures.some(([r]) => /rate|429/i.test(r))) {
    actions.push("🐢  Rate-limiting detected — reduce concurrency and add back-off delays.");
  }
  if (requeueCount > total * 0.3) {
    actions.push("🔁  High requeue rate — check if credential classification is correct.");
  }
  if (actions.length === 0) {
    actions.push("✅  Run looks healthy. Continue monitoring.");
  }

  // Assemble Markdown
  const now = new Date().toISOString().replace("T", " ").split(".")[0] + " UTC";
  const lines = [
    `# Hermes Run Summary — ${now}`,
    "",
    "## Overview",
    "| Metric | Value |",
    "|--------|-------|",
    `| Total events | ${total} |`,
    `| Unique credentials | ${credentialsSeen.size} |`,
    `| Success | ${successCount} |`,
    `| Failed | ${failCount} |`,
    `| Requeued | ${requeueCount} |`,
    `| Success rate | ${successRate.toFixed(1)}% |`,
    `| Credits consumed | ${totalCredits.toFixed(2)} |`,
    "",
    "## Outcome Breakdown",
  ];

  for (const [outcome, count] of outcomes.mostCommon()) {
    lines.push(`- **${outcome}**: ${count}`);
  }

  lines.push("", "## Top 5 Failure Reasons");
  if (topFailures.length > 0) {
    for (let i = 0; i < topFailures.length; i++) {
      const [reason, count] = topFailures[i]!;
      lines.push(`${i + 1}. \`${reason}\` — ${count} occurrence(s)`);
    }
  } else {
    lines.push("_No failures recorded._");
  }

  lines.push("", "## Recommended Actions");
  for (const a of actions) {
    lines.push(`- ${a}`);
  }

  const markdown = lines.join("\n") + "\n";

  // Persist to disk
  ensureReportsDir();
  const ts = new Date()
    .toISOString()
    .replace(/[-:]/g, "")
    .replace("T", "T")
    .split(".")[0];
  const reportPath = path.join(REPORTS_DIR, `report-${ts}.md`);
  try {
    fs.writeFileSync(reportPath, markdown, "utf-8");
  } catch {
    // Non-fatal if reports dir is not writable
  }

  return {
    total,
    uniqueCredentials: credentialsSeen.size,
    successCount,
    failCount,
    requeueCount,
    successRate: Math.round(successRate * 10) / 10,
    totalCredits: Math.round(totalCredits * 100) / 100,
    outcomeBreakdown: new Map(outcomes.entries()),
    topFailures,
    recommendedActions: actions,
    reportPath,
    markdown,
  };
}

File: hermes/reports.py
"""#20 — Hermes Summary Reports

Generates a Markdown summary report from a list of event dicts.
Reports include credential counts, outcome breakdowns, credit
consumption, top failure reasons, and recommended next actions.
Saved to ``hermes/reports/`` with timestamped filenames.
"""
from __future__ import annotations

import datetime
import logging
from collections import Counter
from pathlib import Path

logger = logging.getLogger(__name__)

REPORTS_DIR = Path(__file__).parent / "reports"


def _ensure_reports_dir() -> Path:
    REPORTS_DIR.mkdir(parents=True, exist_ok=True)
    return REPORTS_DIR


def generate_run_summary(events: list[dict]) -> str:
    """Produce a Markdown report summarising *events*.

    Parameters
    ----------
    events:
        List of raw websocket event dicts.  Each is expected to have a
        ``data`` key with at least ``outcome`` and optionally
        ``creditsSpent``, ``credential``, ``error`` fields.

    Returns
    -------
    The Markdown report as a string.  The report is also written to
    ``hermes/reports/report-{timestamp}.md``.
    """
    total = len(events)
    outcomes: Counter[str] = Counter()
    failure_reasons: Counter[str] = Counter()
    credentials_seen: set[str] = set()
    total_credits = 0.0

    for ev in events:
        data = ev.get("data", {})
        outcome = str(data.get("outcome", "unknown"))
        outcomes[outcome] += 1

        cred = data.get("credential") or data.get("email") or data.get("username")
        if cred:
            credentials_seen.add(str(cred))

        total_credits += float(data.get("creditsSpent", 0.0))

        if outcome not in ("success", "requeue"):
            reason = data.get("error") or data.get("message") or outcome
            failure_reasons[str(reason)] += 1

    success_count = outcomes.get("success", 0)
    fail_count = sum(c for o, c in outcomes.items() if o not in ("success", "requeue"))
    requeue_count = outcomes.get("requeue", 0)
    success_rate = (success_count / total * 100) if total else 0.0

    top_failures = failure_reasons.most_common(5)

    # Build recommended actions
    actions: list[str] = []
    if success_rate < 50:
        actions.append("⚠️  Success rate below 50% — investigate top failure reasons urgently.")
    if any("selector" in r.lower() or "dom" in r.lower() for r, _ in top_failures):
        actions.append("🔧  DOM/selector failures detected — review engine.ts locators against latest page screenshots.")
    if any("proxy" in r.lower() or "tunnel" in r.lower() for r, _ in top_failures):
        actions.append("🌐  Infrastructure failures detected — rotate proxy endpoints or check tunnel health.")
    if any("rate" in r.lower() or "429" in r for r, _ in top_failures):
        actions.append("🐢  Rate-limiting detected — reduce concurrency and add back-off delays.")
    if requeue_count > total * 0.3:
        actions.append("🔁  High requeue rate — check if credential classification is correct.")
    if not actions:
        actions.append("✅  Run looks healthy. Continue monitoring.")

    # Assemble Markdown
    now = datetime.datetime.now(datetime.timezone.utc).strftime("%Y-%m-%d %H:%M:%S UTC")
    lines = [
        f"# Hermes Run Summary — {now}",
        "",
        "## Overview",
        f"| Metric | Value |",
        f"|--------|-------|",
        f"| Total events | {total} |",
        f"| Unique credentials | {len(credentials_seen)} |",
        f"| Success | {success_count} |",
        f"| Failed | {fail_count} |",
        f"| Requeued | {requeue_count} |",
        f"| Success rate | {success_rate:.1f}% |",
        f"| Credits consumed | {total_credits:.2f} |",
        "",
        "## Outcome Breakdown",
    ]
    for outcome, count in outcomes.most_common():
        lines.append(f"- **{outcome}**: {count}")

    lines += ["", "## Top 5 Failure Reasons"]
    if top_failures:
        for i, (reason, count) in enumerate(top_failures, 1):
            lines.append(f"{i}. `{reason}` — {count} occurrence(s)")
    else:
        lines.append("_No failures recorded._")

    lines += ["", "## Recommended Actions"]
    for a in actions:
        lines.append(f"- {a}")

    report = "\n".join(lines) + "\n"

    # Persist to disk
    _ensure_reports_dir()
    ts = datetime.datetime.now(datetime.timezone.utc).strftime("%Y%m%dT%H%M%S")
    report_path = REPORTS_DIR / f"report-{ts}.md"
    report_path.write_text(report, encoding="utf-8")
    logger.info("Saved run summary to %s", report_path)

    return report
